lib/reservations.rb

```diff
@@ -0,0 +1,123 @@
+require 'date'
+
+module Hotel
+  class RoomQuantityError < StandardError
+  end
+
+  class Reservations
+    attr_reader :all_rooms, :all_reservations
+
+    def initialize
+      @all_rooms = []
+      create_rooms
+
+      @all_reservations = []
+
+
+    end
+
+    def create_rooms
+      20.times do |i|
+        num = i + 1
+        @all_rooms << Room.new(num)
+      end
+    end
+
+
+    def make_booking(no_of_rooms, start_date, end_date, block: false)
+
+      date_range = DateRange.new(start_date, end_date)
+
+      availability = check_availability(start_date, end_date)
+      check_if_enough(availability, no_of_rooms)
+
+      id = (@all_reservations.length + 1)
+
+      rooms = create_rooms_for_booking(no_of_rooms, availability, block)
+
+      if block == true
+        check_number_for_block(no_of_rooms)
+        block_id = id
+        booking = Block.new(id, rooms, date_range, block_id)
+      else
+        booking = Booking.new(id, rooms, date_range)
+      end
+
+      @all_reservations << booking
+      return booking
+    end
+
+    def create_rooms_for_booking(no_of_rooms, availability, block)
+      rooms = []
+      no_of_rooms.times do |i|
+        room = availability[i]
+        if block == true
+          room.reserved = false
+        else
+          room.reserved = true
```

Comment:
Toggling `reserved` here doesn't take the dates into account. I suspect this means you will run into trouble if you have multiple blocks containing the same rooms for different dates.

Instead, each block should probably keep track of which rooms are reserved and which are still available.